File: Bench.cs
/*
 * Copyright (C) 2014 Katsuya Iida. All rights reserved.
 */

using System.Collections.Generic;
using System;

public class Bench {

    const int TEST_SIZE = 5000 * 100;

    class User {
        public string Name;
        public string Email;
        public IList<string> MessageList = new List<string>();
        public override string ToString() {
            return Name + "<" + Email + ">";
        }
    }

    struct TestData {
        public string FromId;
        public string ToId;
    }

    IDictionary<string, User> idUserMap = new Dictionary<string, User>();
    LinkedList<TestData> testDataList = new LinkedList<TestData>();
    
    string IndexToId(int index) {
        return "id-" + index;
    }
    
    public void Load()
    {
        // No error check for brevity.
        string[] lines = System.IO.File.ReadAllLines(@"userdata.csv");
        for (int index = 0; index < lines.Length; index++)
        {
            string[] record = lines[index].Split(',');
            string name = record[0];
            string email = record[2];
            string id = IndexToId(index);
            idUserMap[id] = new User {
                Name = name,
                Email = email
            };
        }

        var rnd = new Random();
        int numUsers = idUserMap.Count;
        for (int i = 0; i < TEST_SIZE; i++) {
            string fromId = IndexToId(rnd.Next(numUsers));
            string toId = IndexToId(rnd.Next(numUsers));
            TestData data = new TestData {
                FromId = fromId,
                ToId = toId
            };
            testDataList.AddLast(data);
        }
    }

    public void Test() {
        foreach (var data in testDataList) {
            var fromUser = idUserMap[data.FromId];
            var toUser = idUserMap[data.ToId];
            string message = fromUser + " to " + toUser;
            toUser.MessageList.Add(message);
        }
    }

    public void Examine() {
        int numUsers = idUserMap.Count;
        var rnd = new Random();
        for (int i = 0; i < 3; i++) {
            string id = IndexToId(rnd.Next(numUsers));
            var user = idUserMap[id];
            Console.WriteLine("To: " + user);
            foreach(string message in user.MessageList) {
                Console.WriteLine(message);
            }
        }
    }

    public void Run() {
        Console.WriteLine("loading: " + TEST_SIZE + " test items");
        Load();
        long start = Environment.TickCount;
        Console.WriteLine("start");
        for (int i = 0; i < 4; i++) {
            Test();
            long end = Environment.TickCount;
            Console.WriteLine("time: " + (end - start) + "ms");
        }
        // examine();
    }

    static void Main(string[] args)
    {
        new Bench().Run();
    }
}

File: bench.lua
--
-- Copyright (C) 2014 Katsuya Iida. All rights reserved.
--

local TEST_SIZE = 5000 * 100

User = {}
User.__index = User

function User:new(o)
    local o = o or {}
    setmetatable(o, User)
    self.__index = self
    if not o.messageList then
        o.messageList = {}
    end
    return o
end

function User.__tostring(user)
    return user.name .. " <" .. user.email .. ">"
end

function indexToId(index)
    return "id-" .. index
end

local idUserMap = {}
local testDataList = {}
local numUsers = 0

function load()
    local index = 0
    for line in io.lines("userdata.csv") do
        local p1 = line:find(",")
        local p2 = line:find(",", p1 + 1)
        local name = line:sub(1, p1 - 1)
        local email = line:sub(p2 + 1)
        local user = User:new{name = name, email = email}
        local id = indexToId(index)
        idUserMap[id] = user
        index = index + 1
    end

    numUsers = index
    for i = 1, TEST_SIZE do
        local toId = indexToId(math.random(0, numUsers - 1))
        local fromId = indexToId(math.random(0, numUsers - 1))
        local data = { fromId = fromId, toId = toId }
        testDataList[#testDataList + 1] = data
    end
end

function test()
    for i, data in ipairs(testDataList) do
        local fromUser = idUserMap[data.fromId]
        local toUser = idUserMap[data.toId]
        toUser.messageList[#toUser.messageList + 1] = tostring(fromUser) .. " to " .. tostring(toUser)
    end
end

function examine()
    for i = 1, 3 do
        local id = indexToId(math.random(0, numUsers - 1))
        local user = idUserMap[id]
        print("To: " .. tostring(user))
        for i, message in ipairs(user.messageList) do
            print(message)
        end
    end
end

print("loading " .. TEST_SIZE .. " test items")
load()
collectgarbage()
print("start")
local start = os.clock()
for i = 1, 4 do
    test()
    local endTime = os.clock()
    print("time: " .. (endTime - start))
end
-- examine()
